### PR TITLE
feat: add ability in robyn cli to scaffold example programs with various DBs

### LINKED ISSUE #511: Robyn CLI - Add ability to scaffold different kinds of projects
Issue:

As of now, the Robyn CLI tool has limited scaffolding capabilities and does not support a variety of project types. Developers may need to create applications with different backends such as MongoDB, PostgreSQL, or utilizing an ORM like SQLAlchemy. These different needs necessitate the extension of the Robyn CLI to scaffold these types of projects effectively.

Expected Behavior:

When a user invokes the Robyn CLI to scaffold a new project, they should be able to specify the project type based on the intended backend technology or ORM. The resulting scaffolded project should have the appropriate boilerplate code and dependencies set up for the selected type.

Steps to Reproduce:

Run a Robyn CLI command to scaffold a new project (e.g., `python -m robyn --create`).
Notice the lack of options to specify the project type based on different backend technologies or ORMs.
Suggested Solution:

Extend the Robyn CLI to accept a project type as an argument and scaffold the project accordingly. For instance, we could provide options like --mongo, --postgres, or --sqlalchemy to scaffold a MongoDB, PostgreSQL, or SQLAlchemy project, respectively.

### PR DESCRIPTION
**Pull request by bwdq**
Fixes #511 and adds some example code for scaffolding Sqlite, Postgres, Mongo, Prisma, and SQLAlchemy.

### CODE REVIEW

**Comment by sansyrox** [comment]
Hey @bwdq 👋 

Great work 😄 

I will have a thorough look over the weekend but this great work so far!

File: robyn/scaffold/postgres.py
Comment (sansyrox) [inline robyn/scaffold/postgres.py:40]: Do we need these?

**Review by sansyrox — COMMENTED** [review]
Hey @bwdq 👋 

Great work 😄 

Can you please bring the dockerfile in the scaffold folder instead of generating it on the fly?

File: robyn/scaffold/mongo.py
Comment (sansyrox) [inline robyn/scaffold/mongo.py:1]: All these databases would also require certain dependencies. Can we update the docker files accordingly?

File: robyn/__main__.py
Comment (sansyrox) [inline robyn/__main__.py:37]:
```suggestion
    project_dir, docker, project_type = result
```

File: robyn/__main__.py
Comment (sansyrox) [inline robyn/__main__.py:76]: do we need this?

File: robyn/__main__.py
Comment (bwdq) [reply robyn/__main__.py:37]: I tried this and it did not parse correctly so I left it as is

File: robyn/scaffold/mongo.py
Comment (bwdq) [reply robyn/scaffold/mongo.py:1]: I have updated the Docker files for the databases. When I ran docker it errored 'uvloop not found' so uvloop should be a mandatory dependency instead of optional.

File: robyn/__main__.py
Comment (bwdq) [reply robyn/__main__.py]: I think my IDE did that. Outdated by my changing Dockerfile generation.

File: robyn/scaffold/docker/Dockerfile-mongo
Comment (sansyrox) [inline robyn/scaffold/docker/Dockerfile-mongo:1]: Hey @bwdq 👋 

Can we please move them to something like `/scaffold/mongo/Dockerfile` and the code to `/scaffold/mongo/app.py`

Comment (bwdq) [reply robyn/scaffold/docker/Dockerfile-mongo:1]: done

File: robyn/__main__.py
Comment (sansyrox) [reply robyn/__main__.py:37]: @bwdq , what do you mean by this? What error were you getting?

We can return a dictionary instead of a tuple if there is an issue

File: robyn/scaffold/prisma/app.py
Comment (sansyrox) [inline robyn/scaffold/prisma/app.py:35]: @bwdq 👋 

Can you please explain the line here?

Comment (bwdq) [reply robyn/scaffold/prisma/app.py:35]: These are instructions to the user because prisma won't "just work" and requires some setup. I'm not sure what the optimal solution here is. I don't want to execute bash commands silently.

Comment (sansyrox) [reply robyn/scaffold/prisma/app.py:35]: hey @bwdq 👋 

We can add a small readme inside every folder. Doesn't need to be too fancy. Can just contain these commands.

Comment (bwdq) [reply robyn/scaffold/prisma/app.py]: I have moved the python comments to README's.

File: robyn/scaffold/sqlite/app.py
Comment (sansyrox) [inline robyn/scaffold/sqlite/app.py:17]: I believe that we should be including `res` in the final response.

File: robyn/__main__.py
Comment (sansyrox) [inline robyn/__main__.py:70]: I wonder if there is a cleaner way to implement this?

can you store the file_name and file_paths in a dictionary and then try accessing it?

@IdoKendo do you have any thoughts on this?

File: robyn/__main__.py
Comment (sansyrox) [inline robyn/__main__.py:87]: same for this

File: robyn/__main__.py
Comment (IdoKendo) [reply robyn/__main__.py:70]: I would simplify this whole block to something like these 3 lines:
```py
base_path = f"robyn/scaffold/{project_type}"
shutil.copyfile(f"{base_path}/README.md", readme_path)
shutil.copyfile(f"{base_path}/app.py",  app_file_path)
```
Just make sure the `project_type` default is `no-db` 😃 
Can add some guard clauses to be extra safe, but I think `InquirerPy` does that for you out-of-the-box.

Comment (IdoKendo) [reply robyn/__main__.py:70]: Upon further review of the code, can even do it as simple as:
```py
from distutils.dir_util import copy_tree

copy_tree(project_dir, f"robyn/scaffold/{project_type}")
```

**Review by IdoKendo — CHANGES_REQUESTED** [review]
Hey @bwdq 👋 Nice job!
I've added some comments to simplify and improve upon the changes. LMK if there's anything I can help with!

**Comment by sansyrox** [comment]
Thanks for the review @IdoKendo 😄 

I wasn't aware about this feature 

> from distutils.dir_util import copy_tree

**Comment by sansyrox** [comment]
Any updates on this @bwdq ?

File: robyn/scaffold/mongo/Dockerfile
Comment (bwdq) [reply robyn/scaffold/mongo/Dockerfile]: I have updated the base docker image and it would be helpful if you could test it.

File: robyn/__main__.py
Comment (bwdq) [reply robyn/__main__.py]: done

File: robyn/__main__.py
Comment (bwdq) [reply robyn/__main__.py]: This is much simpler but at the time I didn't want to add a new dependency. I've switched to this now.

File: robyn/scaffold/sqlite/app.py
Comment (bwdq) [reply robyn/scaffold/sqlite/app.py]: done

**Comment by bwdq** [comment]
I have updated and simplified the logic in main. I'm still having issues with the robyn resetting the connection when running on docker and some testing would be helpful.

**Comment by sansyrox** [comment]
Hey @bwdq 👋 

> I'm still having issues with the robyn resetting the connection when running on docker and some testing would be helpful.

Can you please explain what you mean here? I don't understand what you mean by "resetting" the connection.

**Comment by bwdq** [comment]
The issue isn't really related to this PR because it's reproducible on the main branch so I've opened a separate issue for it.
https://github.com/sparckles/robyn/issues/621

**Comment by sansyrox** [comment]
Hey @bwdq 👋 

Is the PR up for review?

**Comment by bwdq** [comment]
> Hey @bwdq 👋
> 
> Is the PR up for review?

Yes it is. The docker issue is now fixed.

**Comment by sansyrox** [comment]
Hey @bwdq 👋 

I have two suggestions for this PR
- there are some conflicts. Can we please resolve them?
- can we add some documentation regarding the new change in the `new_docs` folder?

**Review by Mr-Sunglasses — CHANGES_REQUESTED** [review]

File: robyn/scaffold/no-db/Dockerfile
Comment (bwdq) [reply robyn/scaffold/no-db/Dockerfile:1]: python dockerfile because that way I didn't have to install pip
3.11 was the latest stable version
bookworm is the latest stable debian. I tried Alpine and it didn't work

**Review by sansyrox — COMMENTED** [review]
Hey @bwdq 👋 

I was going trying your branch and when I run the command `python -m robyn --create`

I get the following error

```
Please select project type (Mongo/Postgres/Sqlalchemy/Prisma):  Sqlite
Creating a new Robyn project '.'...
Traceback (most recent call last):
  File "/Users/sanskar/.pyenv/versions/3.10.7/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/sanskar/.pyenv/versions/3.10.7/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Users/sanskar/repos/robyn/robyn/__main__.py", line 68, in <module>
    create_robyn_app()
  File "/Users/sanskar/repos/robyn/robyn/__main__.py", line 51, in create_robyn_app
    copy_tree(f"robyn/scaffold/{project_type}/", project_dir)
  File "/Users/sanskar/repos/robyn/venv/lib/python3.10/site-packages/setuptools/_distutils/dir_util.py", line 135, in copy_tree
    raise DistutilsFileError("cannot copy tree '%s': not a directory" % src)
distutils.errors.DistutilsFileError: cannot copy tree 'robyn/scaffold/sqlite/': not a directory
```

And this is happening for every selection. Can you please have a look?

**Review by Mr-Sunglasses — CHANGES_REQUESTED** [review]

**Comment by sansyrox** [comment]
Hey @bwdq 👋 

This PR has been open for a while now. I will be building upon this to get it landed!

**Comment by sansyrox** [comment]
Hey @IdoKendo @Mr-Sunglasses 👋 

I hope all is well 😄 Would it be possible for you folks to have a final review at this PR?

File: robyn/scaffold/sqlalchemy/app.py
Comment (IdoKendo) [inline robyn/scaffold/sqlalchemy/app.py:19]: Maybe it's better to move the sqlalchemy stuff to another module? models.py perhaps?

**Review by IdoKendo — CHANGES_REQUESTED** [review]
Hey @sansyrox 👋 I think it looks good! I would just make one last change with the SQLAlchemy template and I think we can merge this. 🚀

File: robyn/scaffold/postgres/requirements.txt
Comment (Mr-Sunglasses) [inline robyn/scaffold/postgres/requirements.txt:2]: As `pip3 install psycopg2` did not work due to this [reason](https://stackoverflow.com/questions/5420789/how-to-install-psycopg2-with-pip-on-python), so we should change it to `psycopg2-binary` from `psycopg2`


Error when you only run `pip3 install psycopg2`
```
error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [23 lines of output]
      running egg_info
      creating /private/var/folders/p8/gp_5g0jx34vc15g471jmpnpw0000gn/T/pip-pip-egg-info-qzetowdt/psycopg2.egg-info
      writing /private/var/folders/p8/gp_5g0jx34vc15g471jmpnpw0000gn/T/pip-pip-egg-info-qzetowdt/psycopg2.egg-info/PKG-INFO
      writing dependency_links to /private/var/folders/p8/gp_5g0jx34vc15g471jmpnpw0000gn/T/pip-pip-egg-info-qzetowdt/psycopg2.egg-info/dependency_links.txt
      writing top-level names to /private/var/folders/p8/gp_5g0jx34vc15g471jmpnpw0000gn/T/pip-pip-egg-info-qzetowdt/psycopg2.egg-info/top_level.txt
      writing manifest file '/private/var/folders/p8/gp_5g0jx34vc15g471jmpnpw0000gn/T/pip-pip-egg-info-qzetowdt/psycopg2.egg-info/SOURCES.txt'

      Error: pg_config executable not found.

      pg_config is required to build psycopg2 from source.  Please add the directory
      containing pg_config to the $PATH or specify the full executable path with the
      option:

          python setup.py build_ext --pg-config /path/to/pg_config build ...

      or with the pg_config option in 'setup.cfg'.

      If you prefer to avoid building psycopg2 from source, please install the PyPI
      'psycopg2-binary' package instead.

      For further information please check the 'doc/src/install.rst' file (also at
      <https://www.psycopg.org/docs/install.html>).

      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

```

**Review by Mr-Sunglasses — CHANGES_REQUESTED** [review]

File: robyn/__main__.py
Comment (sansyrox) [inline robyn/__main__.py:7]: To fix
```
/home/batman/repos/robyn/robyn/__main__.py:7: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
  from distutils.dir_util import copy_tree
```

**Review by Mr-Sunglasses — COMMENTED** [review]
My two cents is that if we are adding a sqlalchemy template, then it will be a good idea to add a template of sqlalchemy with migrations ( alembic migrations )

WDYT @sansyrox @IdoKendo

**Comment by sansyrox** [comment]
@Mr-Sunglasses , can you please explain a bit more? What would be different?

**Comment by IdoKendo** [comment]
> My two cents is that if we are adding a sqlalchemy template, then it will be a good idea to add a template of sqlalchemy with migrations ( alembic migrations )
> 
> 
> 
> WDYT @sansyrox @IdoKendo 

Not sure we should hold this PR for this. after it's merged I think we can open a separate issue for this.

**Comment by Mr-Sunglasses** [comment]
> > My two cents is that if we are adding a sqlalchemy template, then it will be a good idea to add a template of sqlalchemy with migrations ( alembic migrations )
> > WDYT @sansyrox @IdoKendo
> 
> Not sure we should hold this PR for this. after it's merged I think we can open a separate issue for this.

I agree with this, We should implement it when we migrate to a new repo #659 .

**Comment by Mr-Sunglasses** [comment]
> @Mr-Sunglasses , can you please explain a bit more? What would be different?

@sansyrox I'm taking about sqlalchemy + alembic migration template but we can do after we migrate create_robyn_app to another repo ( #659 )

File: robyn/__main__.py
Comment (Mr-Sunglasses) [inline robyn/__main__.py:60]: I think it is here by mistake, it should not be here 😢 . It'll cause errors.

**Review by Mr-Sunglasses — CHANGES_REQUESTED** [review]

File: robyn/__main__.py
Comment (sansyrox) [reply robyn/__main__.py]: done!

**Comment by sansyrox** [comment]
> > @Mr-Sunglasses , can you please explain a bit more? What would be different?
> 
> @sansyrox I'm taking about sqlalchemy + alembic migration template but we can do after we migrate create_robyn_app to another repo ( #659 )

@Mr-Sunglasses , can you please make an issue for the same?

**Comment by Mr-Sunglasses** [comment]
> > > @Mr-Sunglasses , can you please explain a bit more? What would be different?
> > 
> > 
> > @sansyrox I'm taking about sqlalchemy + alembic migration template but we can do after we migrate create_robyn_app to another repo ( #659 )
> 
> @Mr-Sunglasses , can you please make an issue for the same?

Sure.

**Review by Mr-Sunglasses — APPROVED** [review]
LGTM Great work Team 🙌🏻

**Review by IdoKendo — APPROVED** [review]
LGTM 🔥

**Comment by sansyrox** [comment]
Thank you @IdoKendo and @Mr-Sunglasses 😄 ✨

**Pull request merged**
